cmd/heimdalld/main.go

```diff
@@ -138,6 +184,254 @@ func exportAppStateAndTMValidators(logger log.Logger, db dbm.DB, storeTracer io.
+	cmd.Flags().Uint64(FlagHaltTime, 0, "Minimum block time (in Unix seconds) at which to gracefully halt the chain and shutdown the node")
+	cmd.Flags().String(flagCPUProfile, "", "Enable CPU profiling and write to the provided file")
+
+	// add support for all Tendermint-specific command line options
+	tcmd.AddNodeFlags(cmd)
+	return cmd
+}
+
+func startStandAlone(ctx *server.Context, appCreator server.AppCreator) error {
+	addr := viper.GetString(flagAddress)
+	home := viper.GetString("home")
+	traceWriterFile := viper.GetString(flagTraceStore)
+
+	db, err := openDB(home)
+	if err != nil {
+		return err
+	}
+	traceWriter, err := openTraceWriter(traceWriterFile)
+	if err != nil {
+		return err
+	}
+
+	app := appCreator(ctx.Logger, db, traceWriter)
+
+	svr, err := tserver.NewServer(addr, "socket", app)
+	if err != nil {
+		return fmt.Errorf("error creating listener: %v", err)
+	}
+
+	svr.SetLogger(ctx.Logger.With("module", "abci-server"))
+
+	err = svr.Start()
+	if err != nil {
+		common.Exit(err.Error())
+	}
+
+	common.TrapSignal(ctx.Logger, func() {
+		// cleanup
+		err = svr.Stop()
+		if err != nil {
+			common.Exit(err.Error())
+		}
+	})
+
+	// run forever (the node will not be returned)
+	select {}
+}
+
+func startInProcess(ctx *server.Context, appCreator server.AppCreator, cdc *codec.Codec, startRestServer bool, startBridge bool) (*node.Node, error) {
+	cfg := ctx.Config
+	home := cfg.RootDir
+	traceWriterFile := viper.GetString(flagTraceStore)
+
+	db, err := openDB(home)
+	if err != nil {
+		return nil, err
+	}
+	traceWriter, err := openTraceWriter(traceWriterFile)
+	if err != nil {
+		return nil, err
+	}
+
+	app := appCreator(ctx.Logger, db, traceWriter)
+
+	nodeKey, err := p2p.LoadOrGenNodeKey(cfg.NodeKeyFile())
+	if err != nil {
+		return nil, err
+	}
+
+	server.UpgradeOldPrivValFile(cfg)
+
+	// create & start tendermint node
+	tmNode, err := node.NewNode(
+		cfg,
+		pvm.LoadOrGenFilePV(cfg.PrivValidatorKeyFile(), cfg.PrivValidatorStateFile()),
+		nodeKey,
+		proxy.NewLocalClientCreator(app),
+		node.DefaultGenesisDocProviderFunc(cfg),
+		node.DefaultDBProvider,
+		node.DefaultMetricsProvider(cfg.Instrumentation),
+		ctx.Logger.With("module", "node"),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := tmNode.Start(); err != nil {
+		return nil, err
+	}
+
+	var cpuProfileCleanup func()
+
+	if cpuProfile := viper.GetString(flagCPUProfile); cpuProfile != "" {
+		f, err := os.Create(cpuProfile)
+		if err != nil {
+			return nil, err
+		}
+
+		ctx.Logger.Info("starting CPU profiler", "profile", cpuProfile)
+		if err := pprof.StartCPUProfile(f); err != nil {
+			return nil, err
+		}
+
+		cpuProfileCleanup = func() {
+			ctx.Logger.Info("stopping CPU profiler", "profile", cpuProfile)
+			pprof.StopCPUProfile()
+			f.Close()
+		}
+	}
+
+	// start rest
+	if startRestServer {
+		restCh := make(chan struct{})
+		go func() {
+			_ = restServer.StartRestServer(cdc, hmserver.RegisterRoutes, restCh)
```

Comment:
Could this just be instead something like this:
```
closeCh := make(chan struct{})
go func() {
    _ = restServer.StartRestServer(cdc, hmserver.RegisterRoutes)
    close(closeCh)
}()
<-closeCh
```
I think this change would not affect either the other call to `StartRestServer`.

Reply:
startRestServer function is blocking, so the channel would never be executed, in the example you provided...